src/metabase/query_processor/streaming/xlsx.clj

```diff
@@ -22,61 +230,84 @@
-;; create a bunch of delays for them that will create them if needed and bind this to `*cell-styles*`; we'll use them
-;; as needed in various `set-cell!` method impls.
-
-(def ^:private ^{:arglists '(^String [style-name])} cell-style->format-string
-  "Predefined cell format strings. These are used to get the corresponding predefined cell format number using
-  `BuiltinFormats/getBuiltinFormat`. See `(BuiltinFormats/getAll)` for all predefined formats"
-  {:date     "m/d/yy"
-   :datetime "m/d/yy h:mm"
-   :time     "h:mm:ss"})
-
-(defn- cell-format
-  "Fetch the predefined cell format integer associated with `style-name`."
-  ^Integer [style-name]
-  (or (some-> style-name cell-style->format-string BuiltinFormats/getBuiltinFormat)
-      (throw (ex-info (tru "Invalid cell format")
-                      {:type          qp.error-type/qp
-                       :style-name    style-name
-                       :known-formats cell-style->format-string
-                       :all-formats   (BuiltinFormats/getAll)}))))
-
-(defn- cell-style-delays
-  "Create a map of style name -> delay. This is bound to `*cell-styles*` by `streaming-results-writer`. Dereffing the
-  delay will create the style and add it to the workbook if needed."
-  [^Workbook workbook]
-  (into {} (for [style-name (keys cell-style->format-string)]
-             [style-name (delay
-                           (doto (.createCellStyle workbook)
-                             (.setDataFormat (cell-format style-name))))])))
+(defn- format-string-delay
+  [^Workbook workbook ^DataFormat data-format format-string]
+  (delay
+   (doto (.createCellStyle workbook)
+     (.setDataFormat (. data-format getFormat ^String format-string)))))
+
+(defn- column-style-delays
+  [^Workbook workbook data-format col-settings cols]
+  (into {} (for [col cols]
+             (let [settings-key  (if (:id col)
+                                   {::mb.viz/field-id (:id col)}
+                                   {::mb.viz/column-name (:name col)})
+                   id-or-name    (first (vals settings-key))
+                   settings      (get col-settings settings-key)
+                   semantic-type (:semantic_type col)
+                   format-strings (format-settings->format-strings settings semantic-type)]
+               (when (seq format-strings)
+                 {id-or-name
+                  (map
+                   #(format-string-delay workbook data-format %)
+                   format-strings)})))))
+
+(def ^:private cell-style-delays
+  "Creates a map of column name or id -> delay, or keyword representing default -> delay. This is bound to
+  `*cell-styles*` by `streaming-results-writer`. Dereffing the delay will create the style and add it to
+  the workbook if needed.
+
+  Memoized so that it can be called within write-row! without re-running the logic to convert format settings
+  to format strings."
+  (memoize
```

Comment:
making a note:
unmemoized I have samples from medium-size xlsx dl times of 2.12s 1.71s 1.60s 1.62s
memoized with same file it's 817ms 460ms 528ms 689 ms - so this deffo seems to be critical path